artifact/v3/manifest/manifest.cpp

```diff
@@ -0,0 +1,68 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <artifact/v3/manifest/manifest.hpp>
+
+#include <string>
+#include <sstream>
+
+#include <common/error.hpp>
+
+namespace artifact {
+namespace v3 {
+namespace manifest {
+
+namespace io = mender::common::io;
+namespace error = mender::common::error;
+
+const size_t expected_shasum_length {64};
+
+ExpectedManifest Parse(mender::common::io::Reader &reader) {
+	Manifest m {};
+
+	stringstream ss {};
+
+	io::StreamWriter sw {ss};
+
+	auto err = io::Copy(sw, reader);
+	if (err) {
+		return expected::unexpected(err);
+	}
+
+	string entry_name {};
+	string shasum {};
+
+	while (!ss.eof()) {
+		ss >> shasum;
+		if (shasum.size() != expected_shasum_length) {
+			return expected::unexpected(error::MakeError(
+				error::GenericError,
+				"Manifest sha's need to be of size " + to_string(expected_shasum_length)
+					+ ". Received: " + to_string(shasum.size())));
+		}
+		/* Ignore the leading whitespace in between entries */
+		ss >> std::ws >> entry_name;
+		if (entry_name == "") {
+			return expected::unexpected(error::MakeError(
+				error::GenericError, "No name found for the manifest entry: " + shasum));
```

Comment:
Shouldn't we at least have a generic `ParsingError` for the artifact parsing? `GenericError` is extremely generic, it's when you don't even have a good category.

Reply:
I might make an error class to share in between all the artifact stuff then